### PR TITLE
Use one codegen class for all Nano factories

### PR DESCRIPTION
**Pull request by ejona86**
This reduces the number of classes defined, which reduces memory usage.
It also reduces the number of methods defined, which is important
because of the dex limit.

This should have virtually zero performance degration because the
contiguous switch uses tableswitch bytecode.

ARG_IN_ and ARG_OUT_ are prefixes as otherwise there could be a name
collision with an oddly-named method. Since the other fields are
prefixed with METHOD_ they can't collide.

### CODE REVIEW

File: compiler/src/java_plugin/cpp/java_generator.cpp [hunk 1]
```diff
@@ -109,6 +111,8 @@ static void PrintMethodFields(
       // share the same request or response messages.
       p->Print(
```
Comment (madongfly) [inline compiler/src/java_plugin/cpp/java_generator.cpp]: The TODO above needs update.

File: compiler/src/java_plugin/cpp/java_generator.cpp [hunk 2]
```diff
@@ -81,6 +81,8 @@ static void PrintMethodFields(
   (*vars)["service_name"] = service->name();
   for (int i = 0; i < service->method_count(); ++i) {
     const MethodDescriptor* method = service->method(i);
+    (*vars)["i"] = to_string(2 * i);
```
Comment (madongfly) [inline compiler/src/java_plugin/cpp/java_generator.cpp]: consider a better name? I'm fine with short names, but since we are inside the for loop, `i` is a little confusion.

File: compiler/src/java_plugin/cpp/java_generator.cpp (hunk 1 above)
Comment (ejona86) [inline compiler/src/java_plugin/cpp/java_generator.cpp]: The TODO is still correct. There are two parsers being created for each method and there is no reuse when message types are shared.

File: compiler/src/java_plugin/cpp/java_generator.cpp (hunk 2 above)
Comment (ejona86) [inline compiler/src/java_plugin/cpp/java_generator.cpp]: Renamed to arg_in_id and arg_out_id.

File: compiler/src/java_plugin/cpp/java_generator.cpp (hunk 1 above)
Comment (madongfly) [inline compiler/src/java_plugin/cpp/java_generator.cpp]: I meant the name, it use `Parser` because we did create `Parser` classes, now it should be `NanoFactory`.

**Comment by madongfly** [comment]
@ejona86 LGTM

File: compiler/src/java_plugin/cpp/java_generator.cpp (hunk 1 above)
Comment (ejona86) [inline compiler/src/java_plugin/cpp/java_generator.cpp]: Oh, I see. Done as a separate commit.
